laplace/baselaplace.py

```diff
@@ -1568,8 +1567,1025 @@ def sample(self, n_samples=100, generator=None):
+
+        if self.diagonal_kernel:
+            return [L_diag[i :: self.n_outputs] for i in range(self.n_outputs)]
+        else:
+            return L_diag
+
+    def _build_Sigma_inv(self):
+        """
+        Computes the cholesky decomposition of
+        \[
+            K_{MM} + \Lambda_{MM}^{-1}.
+        \]
+        See See [Improving predictions of Bayesian neural nets via local linearization (Immer et al., 2021)](https://arxiv.org/abs/2008.08400)
+        Equation 15 for more information.
+
+        As the diagonal approximation is performed with \Lambda_{MM} (which is stored in self.L),
+        the code is greatly simplified.
+        """
+        if self.diagonal_kernel:
+            self.Sigma_inv = [
+                torch.linalg.cholesky(
+                    self.gp_kernel_prior_variance * self.K_MM[c]
+                    + torch.diag(
+                        torch.nan_to_num(1.0 / (self._H_factor * lambda_c), posinf=10.0)
+                    )
+                )
+                for c, lambda_c in enumerate(self.L)
+            ]
+        else:
+            self.Sigma_inv = torch.linalg.cholesky(
+                self.gp_kernel_prior_variance * self.K_MM
+                + torch.diag(
+                    torch.nan_to_num(1 / (self._H_factor * self.L), posinf=10.0)
+                )
+            )
+
+    def _get_SoD_data_loader(self, train_loader: DataLoader) -> DataLoader:
+        """
+        Subset-of-Datapoints data loader
+        """
+        return DataLoader(
+            dataset=train_loader.dataset,
+            batch_size=train_loader.batch_size,
+            sampler=SoDSampler(N=len(train_loader.dataset), M=self.M, seed=self.seed),
+            shuffle=False,
+        )
+
+    def fit(self, train_loader, progress_bar=False):
+        """Fit the Laplace approximation of a GP posterior.
+
+        Parameters
+        ----------
+        train_loader : torch.data.utils.DataLoader
+            `train_loader.dataset` needs to be set to access \\(N\\), size of the data set
+            `train_loader.batch_size` needs to be set to access \\(b\\) batch_size
+        """
+
+        X, _ = next(iter(train_loader))
+        with torch.no_grad():
+            self.n_outputs = self.model(X[:1].to(self._device)).shape[-1]
+        setattr(self.model, 'output_size', self.n_outputs)
+        self.batch_size = train_loader.batch_size
+
+        if (
+            self.likelihood == 'regression'
+            and self.n_outputs > 1
+            and self.diagonal_kernel
+        ):
+            warnings.warn(
+                'Using FunctionalLaplace with the diagonal approximation of a GP kernel is not recommended '
+                'in the case of multivariate regression. Predictive variance will likely be overestimated.'
+            )
+
+        self.model.eval()
+
+        N = len(train_loader.dataset)
+        self.n_data = N
+
+        # by default, all training data points are used for GP inference
```

Comment:
I wonder if this is a good default, in terms of user perception, since this will be very slow. Does it make sense to make `M` mandatory (i.e. not allowing for `None`---the user must specify it)?

Reply:
I do not have a strong opinion here. I'll change it to mandatory.